client/app/lib/hooks/router/dynamicNest/utils.ts

```diff
@@ -0,0 +1,12 @@
+import { useMatches } from 'react-router-dom';
+
+export type Promisable<T> = T | Promise<T>;
+
+export type Match = ReturnType<typeof useMatches>[number];
+
+export const isRecord = <K extends string | number | symbol, V>(
+  value: unknown,
+): value is Record<K, V> => typeof value === 'object' && value !== null;
+
+export const isPromise = <T>(value: unknown): value is Promise<T> =>
```

Comment:
While `instanceof Promise` may work in this instant case, it might not necessarily work everywhere, every time. For example, `instanceof Promise` will not work across [realms](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#sec-code-realms). [This answer](https://stackoverflow.com/a/49832343) provides a good example on how realms work in JavaScript implementations.

Furthermore, JavaScript is a duck-typed language. An object that has a `then` method can also _be considered_ a Promise. MDN describes these types as conforming to the [Thenable interface](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#thenables). Thing is, Promises were only recently becoming a part of the native JS specs, so some libraries may use polyfills that are actually Thenables, not instances of `Promise`. That's why the native `Promise.resolve` or `Promise.all` work even for non-Promise instances that are Thenables. [The discussion here](https://stackoverflow.com/a/28133130) may be useful to refer.

It is thus safest across different browser versions and realms to determine if something is a `Promise` this way, i.e., checking if it is a non-`null` object, and has a `then` attribute that is a `function`. Rather, this function is aptly named `isThenable`, but I chose to name it `isPromise` to make the code easier to understand. After all, a `Promise` is a Thenable. Not conversely, though.